File: FrontEnd/StreaminApp/src/app/services/genre.service.ts
import {Injectable} from '@angular/core';
import {HttpClient, HttpParams} from '@angular/common/http';
import {Observable} from 'rxjs';
import {GenreResponse} from '../models/genre/genre-response';
import {GenreRequest} from '../models/genre/genre-request';

@Injectable({
  providedIn: 'root'
})
export class GenreService {
  private uri: string = '/cinewave/api/v1/genres';

  constructor(private httpClient: HttpClient) {
  }

  public createGenre(genreRequest: GenreRequest): Observable<GenreResponse> {
    return this.httpClient.post(this.uri, genreRequest);
  }

  public modifyGenre(id: string, genreRequest: GenreRequest): Observable<GenreResponse> {
    return this.httpClient.put<GenreResponse>(`${this.uri}/${id}`, genreRequest);
  }

  public deleteGenre(id: string): Observable<GenreResponse> {
    return this.httpClient.delete<GenreResponse>(`${this.uri}/${id}`);
  }

  public findPaginatedGenres(page: number, size: number, sort: string[]): Observable<GenreResponse[]> {
    const params = new HttpParams()
      .set('page', page)
      .set('size', size)
      .set('sort', sort.join(','));
    return this.httpClient.get<GenreResponse[]>(`${this.uri}/paginated`, {params: params});
  }

  public findAllGenres(): Observable<GenreResponse[]> {
    return this.httpClient.get<GenreResponse[]>(this.uri);
  }

  public findGenreById(id: string): Observable<GenreResponse> {
    return this.httpClient.get<GenreResponse>(`${this.uri}/${id}`);
  }

  public findPaginatedGenreByName(page: number, size: number, sort: string[], name: string): Observable<GenreResponse[]> {
    const params: HttpParams = new HttpParams()
      .set('page', page)
      .set('size', size)
      .set('sort', sort.join(','))
      .set('name', name);
    return this.httpClient.get<GenreResponse[]>(`${this.uri}/filter`, {params: params});
  }

  public existsGenreByName(name: string): Observable<any> {
    return this.httpClient.get<any>(`${this.uri}/${name}/exist`);
  }
}

File: BackEnd/uploadService/src/main/java/ma/streaming/upload/genre/GenreService.java
package ma.streaming.upload.genre;

import ma.streaming.upload.genre.dto.GenreRequestDto;
import ma.streaming.upload.genre.dto.GenreResponseDto;
import ma.streaming.upload.shared.exceptions.AlreadyExistsException;
import ma.streaming.upload.shared.exceptions.BadRequestException;
import ma.streaming.upload.shared.exceptions.NotFoundException;
import org.springframework.data.domain.Page;
import org.springframework.stereotype.Repository;

import java.util.List;
import java.util.UUID;

@Repository
public interface GenreService {
    GenreResponseDto createGenre(GenreRequestDto genreRequestDto) throws AlreadyExistsException;

    GenreResponseDto modifyGenre(UUID genreId, GenreRequestDto genreRequestDto) throws AlreadyExistsException, NotFoundException;

    GenreResponseDto deleteGenre(UUID genreId) throws NotFoundException, BadRequestException;

    Page<GenreResponseDto> findPaginatedGenres(int page, int size, String[] sort);

    List<GenreResponseDto> findAllGenres();

    GenreResponseDto findGenreById(UUID genreId) throws NotFoundException;

    Page<GenreResponseDto> findPaginatedGenreByName(int page, int size, String[] sort, String name);

    boolean existsGenreByName(String name);
}
